src/github.com/matrix-org/dendrite/roomserver/api/input.go

```diff
@@ -36,7 +40,51 @@ type InputRoomEvent struct {
 	// For example many matrix events forget to reference the m.room.create event even though it is needed for auth.
 	// (since synapse allows this to happen we have to allow it as well.)
 	AuthEventIDs []string
+	// Whether the state is supplied as a list of event IDs or whether it
+	// should be derived from the state at the previous events.
+	HasState bool
```

Comment:
Oh so it's actually valid to have `len(StateEventIDs) == 0` and yet that really is the room state? Comment this please, as it's pretty important and not obvious!